src/modules/flight_mode_manager/tasks/PrecisionLanding/FlightTaskPrecisionLanding.cpp

```diff
@@ -0,0 +1,317 @@
+{
+	bool ret = FlightTask::activate(last_setpoint);
+	_precland_state = PRECLAND_STATE::AUTORTL_CLIMB;
+
+	_position_setpoint = _position;
+	_search_count = 0;
+
+	_initial_yaw = _yaw;
+	_initial_position = _position;
+
+	return ret;
+}
+
+void FlightTaskPrecisionLanding::do_state_transition(PRECLAND_STATE new_state)
+{
+	_initial_yaw = _yaw;
+	_precland_state = new_state;
+	_state_start_time = hrt_absolute_time();
+}
+
+bool FlightTaskPrecisionLanding::inside_acceptance_radius()
+{
+	// TODO: Reuse what FlightTask has...
+	return matrix::Vector3f(_position_setpoint - _position).norm() <= _param_pld_hacc_rad.get();
+}
+
+bool FlightTaskPrecisionLanding::precision_target_available()
+{
+	const bool ever_received = _landing_target_pose.timestamp != 0;
+	const bool timed_out =  hrt_absolute_time() - _landing_target_pose.timestamp > _param_pld_btout.get() * SEC2USEC;
+	return ever_received && !timed_out;
+}
+
+void FlightTaskPrecisionLanding::generate_pos_xy_setpoints()
+{
+	switch (_precland_state) {
+	case PRECLAND_STATE::AUTORTL_CLIMB:
+		_position_setpoint(0) = _initial_position(0);
+		_position_setpoint(1) = _initial_position(1);
+		break;
+
+	case PRECLAND_STATE::AUTORTL_APPROACH:
+	case PRECLAND_STATE::MOVE_TO_SEARCH_ALTITUDE:
+
+		// Horizontal approach to home or precision target
+		if (precision_target_available()) {
+			_position_setpoint(0) = _landing_target_pose.x_abs;
+			_position_setpoint(1) = _landing_target_pose.y_abs;
+
+		} else {
+			_position_setpoint(0) = _sub_home_position.get().x;
+			_position_setpoint(1) = _sub_home_position.get().y;
+		}
+
+		break;
+
+	case PRECLAND_STATE::SEARCHING_TARGET:
+	case PRECLAND_STATE::MOVING_ABOVE_TARGET:
+	case PRECLAND_STATE::LANDING_ON_TARGET:
+		_position_setpoint(0) = _landing_target_pose.x_abs;
+		_position_setpoint(1) = _landing_target_pose.y_abs;
+		break;
+
+	case PRECLAND_STATE::FALLBACK_LAND:
+               // TODO: might be safer to set only once during state change
+		_position_setpoint(0) = _sub_home_position.get().x;
+		_position_setpoint(1) = _sub_home_position.get().y;
+		break;
+	}
+
+}
+
+void FlightTaskPrecisionLanding::generate_pos_z_setpoints()
+{
+	const float search_rel_altitude = _param_pld_srch_alt.get();
+
+	switch (_precland_state) {
+	case PRECLAND_STATE::AUTORTL_CLIMB:
+	case PRECLAND_STATE::AUTORTL_APPROACH:
+		// Horizontal approach to home or precision target
+		_position_setpoint(2) = _sub_home_position.get().z - _param_rtl_return_alt.get();
+		break;
+
+	case PRECLAND_STATE::MOVE_TO_SEARCH_ALTITUDE:
+	case PRECLAND_STATE::SEARCHING_TARGET:
+	case PRECLAND_STATE::MOVING_ABOVE_TARGET:
+
+		// TODO: Should use current altitude when state was entered as the altitude setpoint
+		if (PX4_ISFINITE(_landing_target_pose.z_abs)) {
+			// For safety reasons take whichever reference altitude is higher up (lower in NED)
+			_position_setpoint(2) = math::min(_sub_home_position.get().z, _landing_target_pose.z_abs) - search_rel_altitude;
```

Comment:
Does this account for the case where home/takeoff point is above the landing tag?